LX-53 Upload all the bundle files at once, instead of one at a time.

pomegranited opened the pull request:
**Description:** When transferring a Block to a Blockstore bundle, upload all the related files at once.

This allows Blockstore to create a single BundleVersion for the change.

**JIRA:** SE-281 

**Dependencies:**

* https://github.com/open-craft/blockstore/pull/8 (merged)

**Installation instructions:**

1. Install this branch in the CMS shell.
1. Run the Blockstore service so the CMS shell can access its REST API.
1. Ensure that the URL for the Blockstore is present in `settings.BLOCKSTORE_API_URL` (see [`openedx_blockstore_relay.settings`](https://github.com/open-craft/openedx-blockstore-relay/tree/master/openedx_blockstore_relay/settings)).

**Testing instructions:**

1. Create a Collection in Blockstore, and note its UUID:
    http://localhost:8888/api/v1/collections/
1. From the Studio shell, run the management command to upload a block to the given collection, e.g.:
   ```bash
   ./manage.py cms transfer_to_blockstore \
        --block-key="block-v1:edX+DemoX+Demo_Course+type@course+block@course" \
        --collection-uuid=a530f2a9-2b0c-4d59-8f41-01878a414102 \
        --settings=devstack_docker
   ```
1. Check that Blockstore has created only a single BundleVersion for this upload, e.g.,
   http://localhost:8888/api/v1/bundle_versions/?bundle__collection__uuid=a530f2a9-2b0c-4d59-8f41-01878a414102

**Reviewers:**
- [ ] @symbolist 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed
- [ ] PR author is listed in AUTHORS

pomegranited committed:
Upload all the bundle files at once, instead of one at a time.

This creates a single BundleVersion for the transfer_to_blockstore management
command, instead of creating a BundleVersion each time a file is uploaded.

Includes quality and test fixes.

bradenmacdonald commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -201,15 +214,19 @@ def _upload_olx(self, olx_node, block):
         block_id = block_key.block_id
         block_type = block_key.block_type
         share_block = (block_key == self.block_key)
+<<<<<<< HEAD
```
@pomegranited You've got some stray conflict markers here.

pomegranited replied:
Ach, sorry!  I tried a rebase first and must have botched it.  Fixed now.

bradenmacdonald commented on openedx_blockstore_relay/transfer_data.py:
```diff
@@ -252,19 +268,43 @@ def _post_bundle_file(self, data, path, name=None, content_type='application/oct
+        Collect and queue the static assets paths scraped from the given data string and block .
         """
         for asset in compat.collect_assets_from_text(data, self.block_key.course_key):
+<<<<<<< HEAD
```
And here.

pomegranited replied:
Fixed in 5f21b78 .

symbolist commented:
@pomegranited Overall seems to do the trick! Can test this out tomorrow once the conflict issue is resolved. Also would it be useful to have some tests for the change?

pomegranited commented:
@symbolist  Sorry about those conflict issues!  I'll work on getting the existing tests running and add new tests tomorrow.

symbolist commented:
@pomegranited I just tested this out and the file upload seems to be working fine. 👍 

I did run into one issue though it is out of scope of this PR. When transferring verticals with videos I am getting the following error:

```
root@8b558e15b50f:/edx/app/edxapp/edx-platform# ./manage.py cms transfer_to_blockstore --settings=devstack_docker --verbosity=2 --block-key "block-v1:edX+DemoX+Demo_Course+type@vertical+block@3dc16db8d14842e38324e95d4030b8a0" --collection-uuid "060f1d4d-3193-4617-b78d-7c087edb505b"

  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/edxval/api.py", line 291, in get_video_transcript_data
    return dict(file_name=video_transcript.filename, content=video_transcript.transcript.file.read())
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/fields/files.py", line 51, in _get_file
    self._file = self.storage.open(self.name, 'rb')
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/files/storage.py", line 38, in open
    return self._open(name, mode)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/files/storage.py", line 300, in _open
    return File(open(self.path(name), mode))
IOError: [Errno 2] No such file or directory: u'/edx/var/edxapp/media/video-transcripts/81f76f854e20495a9fbc5671b2b72491.sjson'
```

Do you see it too?

pomegranited committed:
Updates requests requirement

pomegranited commented:
Thanks for your review @symbolist !  I've modified the tests so that they pass again, and added [this assert](https://github.com/open-craft/openedx-blockstore-relay/pull/5/files#diff-8f43f44a8c09dfd1e467d00808e8b361R59) to ensure that the POST is only being called once for a multi-file block transfer.

Could you please check my test changes and ensure they pass for you too?  [Test instructions here](https://github.com/open-craft/openedx-blockstore-relay/tree/master/openedx_blockstore_relay/tests#openedx-blockstore-relay-tests).

> I did run into one issue though it is out of scope of this PR. When transferring verticals with videos I am getting the following error:

Yep 😦  That's the issue I reported at [the bottom of this comment](https://github.com/open-craft/openedx-blockstore-relay/pull/4#issuecomment-429214991): 

> Note: For this to run without error in my docker devstack, I have to set common.lib.xmodule.xmodule.video_module.edxval_api = None, else I get errors:

bradenmacdonald commented:
Hmm, edx-val is working for me. On my docker devstack, there are `.sjson` files in the `/edx/var/edxapp/media/video-transcripts/` directory. 

If you had edx-val forcibly disabled (via `common.lib.xmodule.xmodule.video_module.edxval_api = None`) when originally editing/creating the video module with transcripts in Studio, that could be why. I'd suggest deleting the transcripts and uploading new ones.

symbolist commented:
@pomegranited Great! I just reviewed the last two commits and ran the test suite.

- [x] I tested this by transferring units from a course to Blockstore.
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

@bradenmacdonald Hmm. I haven't made any changes to `common.lib.xmodule.xmodule.video_module.edxval_api`. And I do not see the dir `/edx/var/edxapp/media/video-transcripts/`. Is there any additional config that needs to be done to enabled edxval?

bradenmacdonald commented:
@symbolist Hmm, I don't remember doing any special config. Did you edit a video XBlock in Studio and upload a transcript file for it?

I do remember one thing that confused me is that /edx/var/edxapp/media/video-transcripts/ will be empty in the LMS container even if it's populated in the Studio container - it's a bug of the docker devstack that that directory is not shared.

pomegranited commented:
Thanks @bradenmacdonald !
I reset my change to `common.lib.xmodule.xmodule.video_module.edxval_api`, and was getting the same error about that missing `.json` file when I tried a Course Export from Studio.  So I re-imported the demo course (using the latest [edx-demo-courses/release](https://github.com/edx/edx-demo-course/releases)), and that got rid of the export error in Studio, and with the `transfer_to_blockstore` command.  So I think we're all good here.
CC @symbolist

symbolist commented:
Ah great!

symbolist commented:
> Did you edit a video XBlock in Studio and upload a transcript file for it?

Hmm. Could be.

> I do remember one thing that confused me is that /edx/var/edxapp/media/video-transcripts/ will be empty in the LMS container even if it's populated in the Studio container - it's a bug of the docker devstack that that directory is not shared.

Okay, that is strange. Though I was checking from the studio-container.
